File: qt/multimedia/gen_qaudiobufferinput.cpp
// +build ignore

#include <QAudioBuffer>
#include <QAudioBufferInput>
#include <QAudioFormat>
#include <QMediaCaptureSession>
#include <QMetaObject>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <qaudiobufferinput.h>
#include "gen_qaudiobufferinput.h"
class MiqtVirtualQAudioBufferInput : public virtual QAudioBufferInput {
public:
MiqtVirtualQAudioBufferInput(): QAudioBufferInput() {};
MiqtVirtualQAudioBufferInput(const QAudioFormat& format): QAudioBufferInput(format) {};
MiqtVirtualQAudioBufferInput(QObject* parent): QAudioBufferInput(parent) {};
MiqtVirtualQAudioBufferInput(const QAudioFormat& format, QObject* parent): QAudioBufferInput(format, parent) {};

virtual ~MiqtVirtualQAudioBufferInput() = default;
// cgo.Handle value for overwritten implementation
	intptr_t handle__MetaObject = 0;
// Subclass to allow providing a Go implementation
	virtual const QMetaObject* metaObject() const override {
if (handle__MetaObject == 0) {
return QAudioBufferInput::metaObject();
}
QMetaObject* callback_return_value = miqt_exec_callback_QAudioBufferInput_MetaObject(const_cast<MiqtVirtualQAudioBufferInput*>(this), handle__MetaObject);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	QMetaObject* virtualbase_MetaObject() const {

		return (QMetaObject*) QAudioBufferInput::metaObject();

	}
// cgo.Handle value for overwritten implementation
	intptr_t handle__Metacast = 0;
// Subclass to allow providing a Go implementation
	virtual void* qt_metacast(const char* param1) override {
if (handle__Metacast == 0) {
return QAudioBufferInput::qt_metacast(param1);
}
const char* sigval1 = (const char*) param1;

		void* callback_return_value = miqt_exec_callback_QAudioBufferInput_Metacast(this, handle__Metacast, sigval1);

		return callback_return_value;
	}
// Wrapper to allow calling protected method
	void* virtualbase_Metacast(const char* param1) {

		return QAudioBufferInput::qt_metacast(param1);

	}
};
QAudioBufferInput* QAudioBufferInput_new() {
return new MiqtVirtualQAudioBufferInput();
}
QAudioBufferInput* QAudioBufferInput_new2(QAudioFormat* format) {
return new MiqtVirtualQAudioBufferInput(*format);
}
QAudioBufferInput* QAudioBufferInput_new3(QObject* parent) {
return new MiqtVirtualQAudioBufferInput(parent);
}
QAudioBufferInput* QAudioBufferInput_new4(QAudioFormat* format, QObject* parent) {
return new MiqtVirtualQAudioBufferInput(*format, parent);
}
void QAudioBufferInput_virtbase(QAudioBufferInput* src
, QObject** outptr_QObject
) {
*outptr_QObject = static_cast<QObject*>(src);
}
QMetaObject* QAudioBufferInput_MetaObject(const QAudioBufferInput* self) {
	return (QMetaObject*) self->metaObject();
}
void* QAudioBufferInput_Metacast(QAudioBufferInput* self, const char* param1) {
	return self->qt_metacast(param1);
}
struct miqt_string QAudioBufferInput_Tr(const char* s) {
	QString _ret = QAudioBufferInput::tr(s);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
bool QAudioBufferInput_SendAudioBuffer(QAudioBufferInput* self, QAudioBuffer* audioBuffer) {
	return self->sendAudioBuffer(*audioBuffer);
}
QAudioFormat* QAudioBufferInput_Format(const QAudioBufferInput* self) {
	return new QAudioFormat(self->format());
}
QMediaCaptureSession* QAudioBufferInput_CaptureSession(const QAudioBufferInput* self) {
	return self->captureSession();
}
void QAudioBufferInput_ReadyToSendAudioBuffer(QAudioBufferInput* self) {
	self->readyToSendAudioBuffer();
}
void QAudioBufferInput_connect_ReadyToSendAudioBuffer(QAudioBufferInput* self, intptr_t slot) {
	MiqtVirtualQAudioBufferInput::connect(self, static_cast<void (QAudioBufferInput::*)()>(&QAudioBufferInput::readyToSendAudioBuffer), self, [=]() {
		miqt_exec_callback_QAudioBufferInput_ReadyToSendAudioBuffer(slot);
	});
}
struct miqt_string QAudioBufferInput_Tr2(const char* s, const char* c) {
	QString _ret = QAudioBufferInput::tr(s, c);
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
struct miqt_string QAudioBufferInput_Tr3(const char* s, const char* c, int n) {
	QString _ret = QAudioBufferInput::tr(s, c, static_cast<int>(n));
	// Convert QString from UTF-16 in C++ RAII memory to UTF-8 in manually-managed C memory
	QByteArray _b = _ret.toUtf8();
	struct miqt_string _ms;
	_ms.len = _b.length();
	_ms.data = static_cast<char*>(malloc(_ms.len));
	memcpy(_ms.data, _b.data(), _ms.len);
	return _ms;
}
void QAudioBufferInput_override_virtual_MetaObject(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQAudioBufferInput*>( (QAudioBufferInput*)(self) )->handle__MetaObject = slot;
}
QMetaObject* QAudioBufferInput_virtualbase_MetaObject(const void* self) {
	return ( (const MiqtVirtualQAudioBufferInput*)(self) )->virtualbase_MetaObject();
}
void QAudioBufferInput_override_virtual_Metacast(void* self, intptr_t slot) {
	dynamic_cast<MiqtVirtualQAudioBufferInput*>( (QAudioBufferInput*)(self) )->handle__Metacast = slot;
}
void* QAudioBufferInput_virtualbase_Metacast(void* self, const char* param1) {
	return ( (MiqtVirtualQAudioBufferInput*)(self) )->virtualbase_Metacast(param1);
}
void QAudioBufferInput_Delete(QAudioBufferInput* self, bool isSubclass) {
	if (isSubclass) {
		delete dynamic_cast<MiqtVirtualQAudioBufferInput*>( self );
	} else {
		delete self;
	}
}

File: qt/multimedia/gen_qaudiobufferinput.go
package multimedia

import (
	"unsafe"

	"github.com/mappu/miqt/qt"
)

type QAudioBufferInput struct {
	h          uintptr
	isSubclass bool
}

// NewQAudioBufferInput constructs a new QAudioBufferInput object.
func NewQAudioBufferInput() *QAudioBufferInput {
	ret := newQAudioBufferInput(QAudioBufferInput_new())
	ret.isSubclass = true
	return ret
}

// NewQAudioBufferInput2 constructs a new QAudioBufferInput object.
func NewQAudioBufferInput2(format *QAudioFormat) *QAudioBufferInput {
	ret := newQAudioBufferInput(QAudioBufferInput_new2(format.cPointer()))
	ret.isSubclass = true
	return ret
}

// NewQAudioBufferInput3 constructs a new QAudioBufferInput object.
func NewQAudioBufferInput3(parent *qt.QObject) *QAudioBufferInput {
	ret := newQAudioBufferInput(QAudioBufferInput_new3((*QObject)(parent.UnsafePointer())))
	ret.isSubclass = true
	return ret
}

// NewQAudioBufferInput4 constructs a new QAudioBufferInput object.
func NewQAudioBufferInput4(format *QAudioFormat, parent *qt.QObject) *QAudioBufferInput {
	ret := newQAudioBufferInput(QAudioBufferInput_new4(format.cPointer(), (*QObject)(parent.UnsafePointer())))
	ret.isSubclass = true
	return ret
}

func (this *QAudioBufferInput) MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QAudioBufferInput_MetaObject(this.h)))
}

func (this *QAudioBufferInput) Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))
	return (unsafe.Pointer)(QAudioBufferInput_Metacast(this.h, param1_Cstring))
}

func QAudioBufferInput_Tr(s string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	var _ms struct_miqt_string = QAudioBufferInput_Tr(s_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QAudioBufferInput) SendAudioBuffer(audioBuffer *QAudioBuffer) bool {
	return (bool)(QAudioBufferInput_SendAudioBuffer(this.h, audioBuffer.cPointer()))
}

func (this *QAudioBufferInput) Format() *QAudioFormat {
	_goptr := newQAudioFormat(QAudioBufferInput_Format(this.h))
	_goptr.GoGC() // Qt uses pass-by-value semantics for this type. Mimic with finalizer
	return _goptr
}

func (this *QAudioBufferInput) CaptureSession() *QMediaCaptureSession {
	return newQMediaCaptureSession(QAudioBufferInput_CaptureSession(this.h))
}

func (this *QAudioBufferInput) ReadyToSendAudioBuffer() {
	QAudioBufferInput_ReadyToSendAudioBuffer(this.h)
}

func (this *QAudioBufferInput) OnReadyToSendAudioBuffer(slot func()) {
	QAudioBufferInput_connect_ReadyToSendAudioBuffer(this.h, intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioBufferInput_ReadyToSendAudioBuffer
func miqt_exec_callback_QAudioBufferInput_ReadyToSendAudioBuffer(cb intptr_t) {
	gofunc, ok := cgo.Handle(cb).Value().(func())
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	gofunc()
}

func QAudioBufferInput_Tr2(s string, c string) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QAudioBufferInput_Tr2(s_Cstring, c_Cstring)
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func QAudioBufferInput_Tr3(s string, c string, n int) string {
	s_Cstring := CString(s)
	defer free(unsafe.Pointer(s_Cstring))
	c_Cstring := CString(c)
	defer free(unsafe.Pointer(c_Cstring))
	var _ms struct_miqt_string = QAudioBufferInput_Tr3(s_Cstring, c_Cstring, (int)(n))
	_ret := GoStringN(_ms.data, int(int64(_ms.len)))
	free(unsafe.Pointer(_ms.data))
	return _ret
}

func (this *QAudioBufferInput) callVirtualBase_MetaObject() *qt.QMetaObject {
	return qt.UnsafeNewQMetaObject(unsafe.Pointer(QAudioBufferInput_virtualbase_MetaObject(unsafe.Pointer(this.h))))
}

func (this *QAudioBufferInput) OnMetaObject(slot func(super func() *qt.QMetaObject) *qt.QMetaObject) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QAudioBufferInput_override_virtual_MetaObject(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioBufferInput_MetaObject
func miqt_exec_callback_QAudioBufferInput_MetaObject(self QAudioBufferInput, cb intptr_t) *QMetaObject {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func() *qt.QMetaObject) *qt.QMetaObject)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	virtualReturn := gofunc((&QAudioBufferInput{h: self}).callVirtualBase_MetaObject)

	return (*QMetaObject)(virtualReturn.UnsafePointer())
}

func (this *QAudioBufferInput) callVirtualBase_Metacast(param1 string) unsafe.Pointer {
	param1_Cstring := CString(param1)
	defer free(unsafe.Pointer(param1_Cstring))

	return (unsafe.Pointer)(QAudioBufferInput_virtualbase_Metacast(unsafe.Pointer(this.h), param1_Cstring))
}

func (this *QAudioBufferInput) OnMetacast(slot func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer) {
	if !this.isSubclass {
		panic("miqt: can only override virtual methods for directly constructed types")
	}
	QAudioBufferInput_override_virtual_Metacast(unsafe.Pointer(this.h), intptr_t(cgo.NewHandle(slot)))
}

//export miqt_exec_callback_QAudioBufferInput_Metacast
func miqt_exec_callback_QAudioBufferInput_Metacast(self QAudioBufferInput, cb intptr_t, param1 *const_char) unsafe.Pointer {
	gofunc, ok := cgo.Handle(cb).Value().(func(super func(param1 string) unsafe.Pointer, param1 string) unsafe.Pointer)
	if !ok {
		panic("miqt: callback of non-callback type (heap corruption?)")
	}

	// Convert all CABI parameters to Go parameters
	param1_ret := param1
	slotval1 := GoString(param1_ret)

	virtualReturn := gofunc((&QAudioBufferInput{h: self}).callVirtualBase_Metacast, slotval1)

	return virtualReturn
}
